nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-processors/src/main/java/org/apache/nifi/processors/elasticsearch/PutElasticsearchHttp.java

```diff
@@ -0,0 +1,330 @@
+        }
+
+        OkHttpClient okHttpClient = getClient();
+        final ComponentLog logger = getLogger();
+
+        // Keep track of the list of flow files that need to be transferred. As they are transferred, remove them from the list.
+        List<FlowFile> flowFilesToTransfer = new LinkedList<>(flowFiles);
+        try {
+            final StringBuilder sb = new StringBuilder();
+            final String baseUrl = trimToEmpty(context.getProperty(ES_URL).getValue());
+            final URL url = new URL((baseUrl.endsWith("/") ? baseUrl : baseUrl + "/") + "_bulk");
+
+            for (FlowFile file : flowFiles) {
+                final String index = context.getProperty(INDEX).evaluateAttributeExpressions(file).getValue();
+                if (StringUtils.isEmpty(index)) {
+                    logger.error("No value for index in for {}, transferring to failure", new Object[]{id_attribute, file});
+                    flowFilesToTransfer.remove(file);
+                    session.transfer(file, REL_FAILURE);
+                    continue;
+                }
+                final String docType = context.getProperty(TYPE).evaluateAttributeExpressions(file).getValue();
+                String indexOp = context.getProperty(INDEX_OP).evaluateAttributeExpressions(file).getValue();
+                if (StringUtils.isEmpty(indexOp)) {
+                    logger.error("No Index operation specified for {}, transferring to failure.", new Object[]{file});
+                    flowFilesToTransfer.remove(file);
+                    file = session.penalize(file);
+                    session.transfer(file, REL_FAILURE);
+                    continue;
+                }
+
+                switch (indexOp.toLowerCase()) {
+                    case "index":
+                    case "update":
+                    case "upsert":
+                    case "delete":
+                        break;
+                    default:
+                        logger.error("Index operation {} not supported for {}, transferring to failure.", new Object[]{indexOp, file});
+                        flowFilesToTransfer.remove(file);
+                        file = session.penalize(file);
+                        session.transfer(file, REL_FAILURE);
+                        continue;
+                }
+
+                final String id = (id_attribute != null) ? file.getAttribute(id_attribute) : null;
+
+                // The ID must be valid for all operations except "index". For that case,
+                // a missing ID indicates one is to be auto-generated by Elasticsearch
+                if (id == null && !indexOp.equalsIgnoreCase("index")) {
+                    logger.error("Index operation {} requires a valid identifier value from a flow file attribute, transferring to failure.",
+                            new Object[]{indexOp, file});
+                    flowFilesToTransfer.remove(file);
+                    file = session.penalize(file);
```

Comment:
Same comment as above, will penalizing this flowfile potentially for this reason help it be processed later? If not then it probably shouldn't be.